<a href="https://colab.research.google.com/github/Auvarmf/Artificial-Intelligence/blob/main/2117051027_AuvarMahsaFahlevi_KuisAI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Kuis Artificial Intelligence


---


Membuat klasifikasi kualitas Wine menggunakan model klasifikasi linear dengan minimal akurasi 60%

## Library

In [25]:
import pandas as pd
import numpy as np
import sklearn 
from sklearn.cluster import AgglomerativeClustering
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

## Dataset
Menggunakan dataset csv dari github:

https://raw.githubusercontent.com/adivii/datasets/main/Artificial%20Intelligence/Pertemuan%205/Wine_Quality_Data.csv 

untuk mencari nilai akurasi kualitas pada Wine

In [26]:
# melakukan pengambilan data
df = pd.read_csv("https://raw.githubusercontent.com/adivii/datasets/main/Artificial%20Intelligence/Pertemuan%205/Wine_Quality_Data.csv")
df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1138            6.3             0.510         0.13             2.3      0.076   
1139            6.8             0.620         0.08             1.9      0.068   
1140            6.2             0.600         0.08             2.0      0.090   
1141            5.9             0.550         0.10             2.2      0.062   
1142            5.9             0.645         0.12             2.0      0.075   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1138                 29.0                  40.0  0.99574  3.42       0.75   
1139                 28.0                  38.0  0.99651  3.42       0.82   
1140                 32.0                  44.0  0.99490  3.45       0.58   
1141                 39.0                  51.0  0.99512  3.52       0.76   
1142                 32.0                  44.0  0.99547  3.57       0.71   

      alcohol  quality    Id  
0         9.4        5     0  
1         9.8        5     1  
2         9.8        5     2  
3         9.8        6     3  
4         9.4        5     4  
...       ...      ...   ...  
1138     11.0        6  1592  
1139      9.5        6  1593  
1140     10.5        5  1594  
1141     11.2        6  1595  
1142     10.2        5  1597  

[1143 rows x 13 columns]

## Preprocessing


*   Pada proses ini dilakukan penghapusan data duplikat dan nilai NaN (Jika diperlukan) pada kasus ini tidak diperlukan.
*   Dilakukan juga pemilihan feature untuk membuat klasifikasi kualitas wine





In [27]:
# # menghapus data duplikat
# df = df.drop_duplicates()
# df

In [28]:
# # kemungkinan tidak terdapat nilai NaN pada dataset tersebut, tapi disini saya tetap memasukan fungsi penghapusan untuk nilai NaN
# # berjaga-jaga jika terjadi penambahan data dan terdapat nilai NaN
# # menghilangkan nilai NaN pada dataset
# df = df.dropna()
# df

# pada kasus ini tidak diperlukan

In [29]:
# # melakukan pemilihan feature untuk membuat klasifikasi kualitas wine
# X = df.drop(["quality"], axis=1)
# Y = df["quality"]

# # melakukan standarisasi pada dataset
# scaler = StandardScaler()
# X = scaler.fit_transform(X)

# karena terjadi ValueError: could not convert string to float: 'red'
# hal ini terjadi karena kolom pada dataframe yang berisi string tidak dapat diubah menjadi float. 
# Pada kasus ini, kolom "color" pada dataframe mengandung string "red" dan "white" yang tidak dapat diubah menjadi int.

# error ini dapat diatasi dengan melakukan encoding pada kolom "color" menjadi angka menggunakan label encoding atau one-hot encoding.
# one_hot = pd.get_dummies(df["color"])

# # menggabungkan one-hot encoding dengan dataframe
# df = pd.concat([df, one_hot], axis=1)
# df

In [30]:
# melakukan drop kolom "color"
df.drop("Id", axis=1)

# memisahkan feature dan target
X = df.drop("quality", axis=1)
y = df["quality"]

# mealkukan split/memisahkan menjadi data train dan test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)

# random_state adalah argumen opsional yang digunakan untuk menentukan seed atau biji secara acak yang digunakan dalam pembagian data.
# test_size digunakan untuk menentukan ukuran data test sebagai proporsi dari keseluruhan dataset.
# test_size = 0.2, maka 20% dari data akan digunakan sebagai data test, sementara 80% sisanya akan digunakan sebagai data training.
# ukuran test_size bisa disesuaikan dengan data yang dimiliki.

## Modeling
Modeling yang digunakan adalah model klasifikasi linear, melakukan klasifikasi pada data dengan output yang bersifat biner (bernilai 0 atau 1). 

Metode ini menggunakan fungsi logistik (sigmoid function) untuk menghitung probabilitas kelas dan kemudian menentukan kelas yang paling mungkin berdasarkan nilai probabilitas tersebut.

In [31]:
# menggunakan model klasifikasi linear
model = LogisticRegression(max_iter=1143)

# melatih model data
model.fit(X_train, y_train)

# memprediksi label pada data test
y_prediksi = model.predict(X_test)

/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [32]:
# melakukan perhitungan akurasi pada data
akurasi = accuracy_score(y_test, y_prediksi)
akurasi

0.5633187772925764

In [33]:
# mengubah akurasi data menjadi percent %
akurasi = accuracy_score(y_test, y_prediksi)
akurasi_percent = akurasi * 100
print("Akurasi :", akurasi_percent, "%")

Akurasi : 56.33187772925764 %


Karena akurasi minimal adalah 60%, disini saya akan mencoba menggunakan model lain agar hasinya sesuai dengan yang diharapkan.

In [34]:
# model yang digunakan adalah Random Forest 
# Random Forest Classifier digunakan dengan 300 estimators (jumlah pohon keputusan yang dibangun) dan random state sebesar 2.
model = RandomForestClassifier(n_estimators=300, random_state=2)
model

RandomForestClassifier(n_estimators=300, random_state=2)

In [35]:
# melatih model menggunakan data train
model.fit(X_train, y_train)

# melakukan prediksi pada data test
y_prediksi = model.predict(X_test)
y_prediksi

array([7, 6, 6, 5, 7, 6, 6, 5, 6, 5, 5, 5, 5, 7, 5, 5, 5, 5, 6, 7, 7, 6,
       6, 5, 5, 5, 6, 6, 5, 5, 6, 5, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 5,
       5, 6, 5, 6, 7, 6, 5, 5, 6, 5, 6, 6, 5, 5, 5, 5, 7, 5, 6, 6, 6, 5,
       6, 5, 6, 6, 6, 6, 7, 6, 6, 5, 6, 7, 6, 7, 5, 5, 6, 5, 6, 5, 6, 6,
       6, 5, 5, 6, 5, 6, 5, 6, 5, 7, 5, 5, 5, 7, 5, 5, 6, 5, 6, 5, 6, 5,
       6, 5, 6, 5, 6, 6, 7, 5, 5, 7, 6, 5, 7, 5, 5, 6, 6, 6, 6, 6, 5, 7,
       5, 6, 5, 6, 5, 6, 6, 5, 7, 6, 5, 5, 7, 6, 6, 6, 5, 5, 5, 5, 7, 6,
       5, 6, 5, 7, 5, 5, 5, 5, 5, 6, 5, 5, 6, 5, 5, 5, 6, 7, 6, 6, 7, 6,
       5, 5, 5, 7, 5, 6, 6, 5, 5, 6, 5, 6, 5, 5, 5, 6, 5, 6, 5, 6, 7, 5,
       6, 6, 6, 5, 6, 5, 5, 6, 6, 6, 6, 6, 5, 6, 5, 6, 5, 6, 5, 5, 6, 6,
       6, 5, 6, 5, 5, 6, 5, 5, 5])

In [36]:
# menghitung akurasi pada data test
akurasi = accuracy_score(y_test, y_prediksi)

# mengubah akurasi ke percent
akurasi_percent = akurasi * 100
print("Akurasi :", akurasi_percent, "%")

Akurasi : 67.24890829694323 %


Nilai akurasi yang bisa saya dapatkan setelah menggunakan Random Forest Classifier adalah 67.24890829694323 %